pkg-r/R/DataSource.R

```diff
@@ -378,6 +373,182 @@ DBISource <- R6::R6Class(
 )
 
 
+#' Data Source: Lazy Tibble
+#'
+#' @description
+#' A DataSource implementation for lazy tibbles connected to databases via
+#' [dbplyr::tbl_sql()] or [dplyr::sql()].
+#'
+#' @examplesIf rlang::is_interactive() && rlang::is_installed("dbplyr") && rlang::is_installed("dplyr") && rlang::is_installed("duckdb")
+#' con <- DBI::dbConnect(duckdb::duckdb())
+#' DBI::dbWriteTable(con, "mtcars", mtcars)
+#'
+#' mtcars_source <- TblLazySource$new(tbl(con, "mtcars"))
+#' mtcars_source$get_db_type()  # "DuckDB"
+#'
+#' result <- mtcars_source$execute_query("SELECT * FROM mtcars WHERE cyl > 4")
+#'
+#' # Note, the result is not the *full* data frame, but a lazy SQL tibble
+#' result
+#'
+#' # You can chain this result into a dplyr pipeline
+#' dplyr::count(result, cyl, gear)
+#'
+#' # Or collect the entire data frame into local memory
+#' dplyr::collect(result)
+#'
+#' # Finally, clean up when done with the database (closes the DB connection)
+#' mtcars_source$cleanup()
+#'
+#' @export
+TblLazySource <- R6::R6Class(
+  "TblLazySource",
+  inherit = DBISource,
+  private = list(
+    tbl = NULL,
+    tbl_cte = NULL
+  ),
+  public = list(
+    #' @field table_name Name of the table to be used in SQL queries
+    table_name = NULL,
+
+    #' @description
+    #' Create a new TblLazySource
+    #'
+    #' @param tbl A [dbplyr::tbl_sql()] (or lazy tibble via [dplyr::tbl()]).
+    #' @param table_name Name of the table in the database. Can be a character
+    #'   string, or will be inferred from the `tbl` argument, if possible.
+    #' @return A new TblLazySource object
+    #' @examplesIf rlang::is_interactive() && rlang::is_installed("dbplyr") && rlang::is_installed("dplyr") && rlang::is_installed("RSQLite")
+    #' conn <- DBI::dbConnect(RSQLite::SQLite(), ":memory:")
+    #' DBI::dbWriteTable(conn, "mtcars", mtcars)
+    #' source <- TblLazySource$new(dplyr::tbl(con, "mtcars"))
+    initialize = function(tbl, table_name = missing_arg()) {
+      check_installed("dbplyr")
+      check_installed("dplyr")
+
+      if (!inherits(tbl, "tbl_sql")) {
+        cli::cli_abort(
+          "{.arg tbl} must be a lazy tibble connected to a database, not {.obj_type_friendly {tbl}}"
+        )
+      }
+
+      private$conn <- dbplyr::remote_con(tbl)
+      private$tbl <- tbl
+
+      # Collect various signals to infer the table name
+      obj_name <- deparse1(substitute(tbl))
```

Comment:
I think we should be making `table_name` required and keep any `substitute()` magic further up in the call stack.

Somewhat relatedly, do you see any utility to exporting the `DataSource` implementations (i.e., `DataFrameSource`, etc)?